File: tools/thrift/lib/go/thrift/exception_test.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package thrift

import (
	"errors"
	"testing"
)

func TestPrependError(t *testing.T) {
	err := NewTApplicationException(INTERNAL_ERROR, "original error")
	err2, ok := PrependError("Prepend: ", err).(TApplicationException)
	if !ok {
		t.Fatal("Couldn't cast error TApplicationException")
	}
	if err2.Error() != "Prepend: original error" {
		t.Fatal("Unexpected error string")
	}
	if err2.TypeId() != INTERNAL_ERROR {
		t.Fatal("Unexpected type error")
	}

	err3 := NewTProtocolExceptionWithType(INVALID_DATA, errors.New("original error"))
	err4, ok := PrependError("Prepend: ", err3).(TProtocolException)
	if !ok {
		t.Fatal("Couldn't cast error TProtocolException")
	}
	if err4.Error() != "Prepend: original error" {
		t.Fatal("Unexpected error string")
	}
	if err4.TypeId() != INVALID_DATA {
		t.Fatal("Unexpected type error")
	}

	err5 := NewTTransportException(TIMED_OUT, "original error")
	err6, ok := PrependError("Prepend: ", err5).(TTransportException)
	if !ok {
		t.Fatal("Couldn't cast error TTransportException")
	}
	if err6.Error() != "Prepend: original error" {
		t.Fatal("Unexpected error string")
	}
	if err6.TypeId() != TIMED_OUT {
		t.Fatal("Unexpected type error")
	}

	err7 := errors.New("original error")
	err8 := PrependError("Prepend: ", err7)
	if err8.Error() != "Prepend: original error" {
		t.Fatal("Unexpected error string")
	}
}

File: test/cpp/end2end/exception_test.cc
/*
 *
 * Copyright 2017 gRPC authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include <exception>
#include <memory>

#include <grpc/impl/codegen/port_platform.h>
#include <grpcpp/channel.h>
#include <grpcpp/client_context.h>
#include <grpcpp/server.h>
#include <grpcpp/server_builder.h>
#include <grpcpp/server_context.h>

#include "src/proto/grpc/testing/echo.grpc.pb.h"
#include "test/core/util/test_config.h"

#include <gtest/gtest.h>

namespace grpc {
namespace testing {

const char* kErrorMessage = "This service caused an exception";

#if GRPC_ALLOW_EXCEPTIONS
class ExceptingServiceImpl : public ::grpc::testing::EchoTestService::Service {
 public:
  Status Echo(ServerContext* /*server_context*/, const EchoRequest* /*request*/,
              EchoResponse* /*response*/) override {
    throw -1;
  }
  Status RequestStream(ServerContext* /*context*/,
                       ServerReader<EchoRequest>* /*reader*/,
                       EchoResponse* /*response*/) override {
    throw ServiceException();
  }

 private:
  class ServiceException final : public std::exception {
   public:
    ServiceException() {}

   private:
    const char* what() const noexcept override { return kErrorMessage; }
  };
};

class ExceptionTest : public ::testing::Test {
 protected:
  ExceptionTest() {}

  void SetUp() override {
    ServerBuilder builder;
    builder.RegisterService(&service_);
    server_ = builder.BuildAndStart();
  }

  void TearDown() override { server_->Shutdown(); }

  void ResetStub() {
    channel_ = server_->InProcessChannel(ChannelArguments());
    stub_ = grpc::testing::EchoTestService::NewStub(channel_);
  }

  std::shared_ptr<Channel> channel_;
  std::unique_ptr<grpc::testing::EchoTestService::Stub> stub_;
  std::unique_ptr<Server> server_;
  ExceptingServiceImpl service_;
};

TEST_F(ExceptionTest, Unary) {
  ResetStub();
  EchoRequest request;
  EchoResponse response;
  request.set_message("test");

  for (int i = 0; i < 10; i++) {
    ClientContext context;
    Status s = stub_->Echo(&context, request, &response);
    EXPECT_FALSE(s.ok());
    EXPECT_EQ(s.error_code(), StatusCode::UNKNOWN);
  }
}

TEST_F(ExceptionTest, RequestStream) {
  ResetStub();
  EchoResponse response;

  for (int i = 0; i < 10; i++) {
    ClientContext context;
    auto stream = stub_->RequestStream(&context, &response);
    stream->WritesDone();
    Status s = stream->Finish();

    EXPECT_FALSE(s.ok());
    EXPECT_EQ(s.error_code(), StatusCode::UNKNOWN);
  }
}

#endif  // GRPC_ALLOW_EXCEPTIONS

}  // namespace testing
}  // namespace grpc

int main(int argc, char** argv) {
  grpc::testing::TestEnvironment env(argc, argv);
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
